### PR TITLE
dev/membership#28 Fix pay-later membership renewal for more than one term

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Alternate to https://github.com/civicrm/civicrm-core/pull/18582

Before
----------------------------------------
Per
https://github.com/civicrm/civicrm-core/pull/18582

After
----------------------------------------
Per
https://github.com/civicrm/civicrm-core/pull/18582

Technical Details
----------------------------------------
@jamie-tillman I've taken your fix & come up with a slightly different take. The difference is mostly that the line items are calculated correctly on the MembershipRenewal form using the BAO_Order object, which is fixed to handle this. The BAO_Order object is a new-ish object, currently only used on that form, intended to provide cleaner internal civicrm methods for calculating line items 

I also switched up the getNumTermsByContributionAndMembershipType slightly - mostly because the CONVERT seemed a bit weird. I think test cover is good enough that we'll get fails if it's not safe to assume qty / num_membership_terms if non-empty - but will see what Jenkins says ....

Comments
----------------------------------------

Adds test cover

### CODE REVIEW

**Comment by jamie-tillman** [comment]
I appreciate you putting the time in to develop this fix further.  Your approach is much better.  I wasn’t familiar yet with the Order API so this is very eye opening for me.  I’ll withdraw my PR and look forward to having your fix in the codebase.

**Comment by eileenmcnaughton** [comment]
Per my notes - this BAO_Order object isn't the order api yet - it's hoped it will be used by it in time but for now it's an internal helper class

**Comment by jamie-tillman** [comment]
@eileenmcnaughton - I went to the test site list linked from the PR and chose "core-18618-8vaf2.test-1.civicrm.org:8001" and ran it through the paces:

* Configure "Fixed" membership, 1 year duration, $1 fee
* Set up a member with that membership who expired in 2019
* Renew for 2 periods and set pending contribution (it auto-calculated $2 fee)
* Go to contributions and complete transaction
* Check membership expiration date - expiration date remained at 12/31/2019

I wonder if I've done something wrong, because it didn't work at all.  In fact, it was worse than before because the membership didn't even advance one period, let alone 2.  It just stayed expired at 12/31/2019.

If you want to view the record I was playing with:  http://core-18618-8vaf2.test-1.civicrm.org:8001/civicrm/contact/view?reset=1&cid=36&key=d334d0e4d9b8bb6ea52e0ff4504aa5b10519f8b0b11874e1ab63c39b5e301a2e_9380

I did this several times and experienced the same result.  However, in one instance I made the transaction complete and not pending and the membership rolled forward 2 periods.

**Comment by jamie-tillman** [comment]
As a follow-up, I noticed that the test site for my PR was still available, so I ran the same sequence on it.  The period rolls forward properly, even with a pending contribution.  Seems something is truly not working with this version of the fix.

http://core-18582-8q6bi.test-1.civicrm.org:8001/civicrm/contact/view?reset=1&cid=36&key=7601518ab4220aa62a92eae948748c9d7b707e6e0c02f276c76c15a2a97863ca_8592&context=search

**Comment by eileenmcnaughton** [comment]
@jamie-tillman arg - I just realised I hadn't pushed up my latest - with https://github.com/civicrm/civicrm-core/pull/18618/files#diff-7a5b0e2d131dabc49178460fc63328c0R441 - I suspect that's the issue (also the cause of the test fails) but I'll re-test locally too.

It will take more than an hour for the new test site to be reliable (it gets built straight away but with the tests still running they can mess with it)

**Comment by eileenmcnaughton** [comment]
@jamie-tillman from my testing it now works when doing the membership renewal & then using 'add payment' but not when updating a contribution to completed - this latter code path goes through a lot of deprecated code. If you can confirm it fixes via 'Payment.create' method & is not worse for the other I think we should merge this & I'll dig a bit more on the other

**Comment by eileenmcnaughton** [comment]
related fix I spotted in testing https://github.com/civicrm/civicrm-core/pull/18620

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Member/Form/MembershipRenewal.php [hunk 1]
```diff
@@ -610,22 +610,13 @@ protected function submit() {
 
     if (!empty($this->_params['record_contribution']) || $this->_mode) {
       // set the source
-      list($userName) = CRM_Contact_BAO_Contact_Location::getEmailDetails(CRM_Core_Session::singleton()->get('userID'));
+      [$userName] = CRM_Contact_BAO_Contact_Location::getEmailDetails(CRM_Core_Session::singleton()->get('userID'));
```
Comment (seamuslee001) [inline CRM/Member/Form/MembershipRenewal.php:613]: @eileenmcnaughton is this the same?

Comment (eileenmcnaughton) [reply CRM/Member/Form/MembershipRenewal.php:613]: @seamuslee001 yep - my ide seems to keep doing it as the preferred new way

**Comment by jamie-tillman** [comment]
Strangely, using my bullet-list test sequence above results in the membership being advanced one too many terms if it's expired.  A membership which expired on 12/31/2018 and is renewed for 2 terms using my test, for example, ends up with a term expiring on 12/31/2021, which is a 3 term renewal.

Yet, if the membership is current (rather than expired), my sequence seems to advance it for the proper number of terms.

Updating the contribution to "Complete" instead of recording payment only advances 1 term in my testing.  If I "edit" the Contribution and actually record the payment while updating, it still advances multiple terms.  

So, it seems the presence of the payment record (or perhaps the payment recording action itself) is what is triggering the multiple terms, rather than the simple completion of the contribution.

--

I think I've figured out why the expired membership is getting the extra term:  The behavior of renewing an expired membership has changed.  If I simply renew a membership that expired on 12/31/2017 and record the Completed Contribution in the renewal form using defaults, the membership is renewed to expire on 12/31/2020 (current).

This is not the behavior I observe on 5.29.1 and prior.  Renewing an expired membership rolls it forward 1 term, even if doing so leaves it expired.

Honestly, if the payment amount could be automatically updated to reflect the price of the multiple terms being applied to membership, this new behavior would be preferable for my clients!  The fact that it charges for 1 term while advanced multiple is the only remaining problem.  If that price calculation were fixed, then renewing a membership would truly be renewing a membership, and no one would have to calculate how many terms are necessary to get it right.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by eileenmcnaughton** [comment]
@jamie-tillman I just pushed a patch that should make the edit contribution flow the same as record payment. However, to the 'what should happen' question....

I just tested on 5.29 and when renewing with payment status = Completed I believe the behaviour is the same with this PR between 5.29 record as completed and this PR record as pending and then complete. The problem is it's confusing



Scenario

- Membership from 17 Jan 2018 to 31 Dec 2018, expired.

If I renew it effective today on 5.29 the membership is updated so that the start date is moved forwards to the beginning of this year and the end date to the end of next year - a 2 year membership consisting of 2 years from the roll over day. However, it's not a contiguous membership - the start date has been altered. This is the same if I record a pending payment & update next week.

So this PR would make it consistent & 'correct' by the current model - it just doesn't meet all your needs.  I see this message - so there IS a confusing way to mess with the selected date for completed and I think that is lost (with & without this change ) for pending + later pay

Screen Shot 2020-10-03 at 3 30 23 PM

My current inclination is to merge this PR  - and then look at some more UI options eg a message to say something like

'This Membership is currently expired. If you renew it the join date will remain the same but the start date will be updated to 1 Jan 2020 to reflect the period they were lapsed for. If you wish to charge them the interim period then you will need to manually adjust the amount field'

I also wonder about providing some options on the record Payment form when the contribution being paid is linked with a membership

ping @pradpnayak @monishdeb @agh1 @Stoob @MegaphoneJon @lcdservices

**Comment by jamie-tillman** [comment]
> 
> My current inclination is to merge this PR - and then look at some more UI options eg a message to say something like
> 
> 'This Membership is currently expired. If you renew it the join date will remain the same but the start date will be updated to 1 Jan 2020 to reflect the period they were lapsed for. If you wish to charge them the interim period then you will need to manually adjust the amount field'
> 
> I also wonder about providing some options on the record Payment form when the contribution being paid is linked with a membership


> 
> 
> @jamie-tillman I just pushed a patch that should make the edit contribution flow the same as record payment. However, to the 'what should happen' question....
> 
> I just tested on 5.29 and when renewing with payment status = Completed I believe the behaviour is the same with this PR between 5.29 record as completed and this PR record as pending and then complete. The problem is it's confusing
> 
> Scenario
> 
>     * Membership from 17 Jan 2018 to 31 Dec 2018, expired.
> 
> 
> If I renew it effective today on 5.29 the membership is updated so that the start date is moved forwards to the beginning of this year and the end date to the end of next year - a 2 year membership consisting of 2 years from the roll over day. However, it's not a contiguous membership - the start date has been altered. This is the same if I record a pending payment & update next week.
> 
> So this PR would make it consistent & 'correct' by the current model - it just doesn't meet all your needs. I see this message - so there IS a confusing way to mess with the selected date for completed and I think that is lost (with & without this change ) for pending + later pay
> 
Screen Shot 2020-10-03 at 3 30 23 PM

As you imply, this message makes it sound like it used to be possible to arrange what my client's current configuration does (figuring the start date from date of expiration, rather than renewal date) but now Civi doesn't appear to do that.  I don't recall modifying any settings that would have enabled this behavior, but it was a couple years ago, so I'll need to investigate the client's config to be sure.

> My current inclination is to merge this PR - and then look at some more UI options eg a message to say something like
> 
> 'This Membership is currently expired. If you renew it the join date will remain the same but the start date will be updated to 1 Jan 2020 to reflect the period they were lapsed for. If you wish to charge them the interim period then you will need to manually adjust the amount field'
> 
> I also wonder about providing some options on the record Payment form when the contribution being paid is linked with a membership

That sounds like an ok approach to me as long as the aforementioned loss/change of functionality is communicated to the user community -- whether caused by this patch or not, it might be a serious issue for some if I'm not the only one in this boat.  I'll have to investigate how to handle this sudden change in functionality for my client.  Perhaps we could just address it with "people processes."

**Comment by eileenmcnaughton** [comment]
I'd like to see some broader engagement on this but part of what I'm seeing is that there is a very blunt way to manipulate how things are processed actually on the renewal form - but that is not retained to be used when the payment completes it.

At the end of the day this is a manual renewal form & we want to give some more control to the person setting up the renewal. I'm pondering whether moving expired memberships into a  'renewal pending' status when the payment is pending & giving the form filler some direct control over the dates makes sense (the status would then result in no-change-in-dates, only status change on completion)

I'll have to think about this some more & then raise a gitlab with the goal of requesting more feedback via this month's dev digest

**Comment by lcdservices** [comment]
This has been a pain point for some time, but without a clear direction on how it should be handled -- as different organizations want to handle the scenario in different ways.

As far as I recall, the current behavior as described above is what we've had in place from early Civi days -- when a renewal is triggered the join date remains the same and the start date is reset to whatever the value would be as if the person was joining now, and the end date configured accordingly. So basically, any "gap" is ignored. If my membership expired in 2018 and I renewed today, it would set the start date without consideration of when I last expired.

I'm not crazy about that. I just don't know that we'll find agreement on how to handle it in a universally accepted way. I think fixing the term qty issue and relying on messaging is acceptable for the addressing the immediate need.

FWIW -- I often advise clients to create a new membership if there is a gap so that the history of the gap is preserved. There are drawbacks to that, and renewals via the frontend won't respect the behavior automatically without customization. But it appeals to me as the best way to capture what's happening.

**Comment by jamie-tillman** [comment]
@lcdservices - Brian, while this makes perfect sense as stated, I am very confused by it, because both of my membership-tracking client sites display the other behavior (roll forward by one period from prior expiration date, even if that period is also expired), and I could swear I've not done anything to adjust the behavior of renewals.

I guess I need to look more closely at the configuration options and at that warning statement that Eileen highlighted, so I can know what to expect going forward and how I can keep things working the way my clients want.

**Comment by agh1** [comment]
I'm concerned about overwriting the line item quantity with the number of terms.  This is mostly a theoretical problem for now since price sets don't allow text/numeric fields to correspond to memberships.  However, I think this could make the situation more confusing to maintain and potentially block off certain use cases.

A common use case is to have a single membership type and offer discounts for buying multiple terms at once.  A price set might have a radio field with $30 for one year and $50 for two years, for example.  *Theoretically,* someone might get `qty` of 3, `membership_num_terms` of 2 because they want to pay $150 to get 6 years of membership.  While it isn't the end of the world if this edge case isn't possible, I don't see why we would want to go out of our way to muddy the distinction between `qty` and `membership_num_terms`.

**Comment by jamie-tillman** [comment]
> 
> 
> I'm concerned about overwriting the line item quantity with the number of terms. This is mostly a theoretical problem for now since price sets don't allow text/numeric fields to correspond to memberships. However, I think this could make the situation more confusing to maintain and potentially block off certain use cases.

I am concerned, as well, and I've already seen an issue with this on printed invoices where the splitting the total amount across the qty value creates strange line item amounts.  But without a facility for tracking `membership_num_terms` beyond the current page transaction (which there is none right now), `qty` is the only available option for pending transactions to complete the task.  

> A common use case is to have a single membership type and offer discounts for buying multiple terms at once. A price set might have a radio field with $30 for one year and $50 for two years, for example. _Theoretically,_ someone might get `qty` of 3, `membership_num_terms` of 2 because they want to pay $150 to get 6 years of membership. While it isn't the end of the world if this edge case isn't possible, I don't see why we would want to go out of our way to muddy the distinction between `qty` and `membership_num_terms`.

To me, it seems this muddiness is already there, by virtue of having implemented Pending Transactions without the necessary data storage to support their implicit functions.  Relying on the membership configuration to know how many terms to advance for a Pending Contribution, as the system did before this PR, has definitely proven to be inadequate.

That said, beyond creating something additional, like a "Pending Renewals" concept, it seems impossible to implement the needed support without using the `qty` field.  I suppose an alternative would be to could leak the `membership_num_terms` concept into the line items table by adding a new column, which would be a hack, but would work to keep `qty` unaffected.

**Comment by agh1** [comment]
> But without a facility for tracking `membership_num_terms` beyond the current page transaction (which there is none right now), `qty` is the only available option for pending transactions to complete the task.

@jamie-tillman your analysis is right on.  I just think it would be less disruptive to add a new mechanism for tracking the number of terms to add rather than shoehorning that into the existing limited schema.

I wonder whether an appropriate place might be, oddly enough, the `civicrm_membership_payment` table.  It has one row per membership-contribution pair, so with an extra `membership_num_terms` field it would be ready to indicate that when contribution `x` is completed, it should increment membership `y` by a specified number of terms.

**Comment by jamie-tillman** [comment]
@agh1, I agree completely.  I'm fairly new at Civi core stuff, so the core behaviors and history are still something I'm coming up to speed on.  But from my limited understanding, your idea sounds great.

**Comment by eileenmcnaughton** [comment]
Just noting that I want to take this back to gitlab & clarify a proposal there - I'm still mulling it - fundamentally the order + later add payment flow does not have a way to track intention. When doing a real-time renewal you have the ability to mess with civi's calculated dates a little - with the blunt instrument of setting the 'renewal date'. I feel like you should be able to indicate more clearly what that renewal should look like.

My proposal was looking to be something like
1) if the status is expired then the renewal would alter it to a new reserved status of 'Pending renewal' with the dates that will be used once the payment is received
2) when the payment is received the status, but not the dates, would be altered.
3) we would expose on the form a bit more info about what dates have been calculated & give the option to specify then on the form
4) we would use the status-override-end-date field to indicate when he status will last until (eg if not paid within a month ) and use the data in the membership_log table to rollback to expired if that date passes.

However, I had assumed, like @jamie-tillman that using qty * membership_num_terms was already part of the model. I think we'd want to test opinions on that more widely before finalising an approach. If we do add a membership_num_terms I would probably put it in the line_items table rather than membership_payment (for legacy reasons we link the contribution to the membership in both places).

I have been thinking about trying to do a 5.32 push on membership issues - with this leading the charge but there are few others

**Pull request closed without merging**